src/support/string.cpp

```diff
@@ -164,44 +164,58 @@ std::ostream& printEscapedJSON(std::ostream& os, const std::string_view str) {
         os << "\\\\";
         continue;
       default: {
+        // Emit something like \u006e, the JSON escaping of a 16-bit number.
         auto uEscape = [&](uint32_t v) {
+          if (v > 0xffff) {
+            std::cerr << "warning: Bad 16-bit escapee " << int(u0) << '\n';
+          }
           os << std::hex;
           os << "\\u";
-          os << ((v >> 24) & 0xff);
-          os << ((v >> 16) & 0xff);
-          os << ((v >> 8) & 0xff);
-          os << (v & 0xff);
+          os << ((v >> 12) & 0xf);
+          os << ((v >> 8) & 0xf);
+          os << ((v >> 4) & 0xf);
+          os << (v & 0xf);
           os << std::dec;
         };
 
         // Based off of
         // https://github.com/emscripten-core/emscripten/blob/59e6b8f1262d75585d8416b728e8cbb3db176fe2/src/library_strings.js#L72-L91
         if (!(u0 & 0x80)) {
           if (u0 >= 32 && u0 < 127) {
+            // This requires no escaping at all.
             os << char(u0);
           } else {
             uEscape(u0);
           }
           continue;
         }
+
+        // This uses 2 bytes.
         i++;
         int u1 = str[i] & 63;
         if ((u0 & 0xE0) == 0xC0) {
           uEscape((((u0 & 31) << 6) | u1));
           continue;
         }
+
+        // This uses 3 bytes.
         i++;
         int u2 = str[i] & 63;
         if ((u0 & 0xF0) == 0xE0) {
           u0 = ((u0 & 15) << 12) | (u1 << 6) | u2;
         } else {
+          // This uses 4 bytes.
+          if ((u0 & 0xF8) != 0xF0) {
+            std::cerr << "warning: Bad UTF-8 leading byte " << int(u0) << '\n';
+          }
```

Comment:
It would be good to emit similar warnings in the 1-, 2-, and 3-byte cases as well